README.txt
==================================
np_trancevibe Max/MSP External
==================================

by Kyle Machulis <kyle@nonpolynomial.com>
Nonpolynomial Labs - http://www.nonpolynomial.com

===========
Description
===========

np_trancevibe is an external for either Max/MSP or Puredata to control the Rez Trancevibrator. It is based on libtrancevibe:

http://sourceforge.net/projects/libtrancevibe

nplabs.trancevibe works with both the original ASCII Rez Trancevibrator, as well as the Drmn Trance Vibrator (http://tim.cexx.org/projects/vibe/)

Portability of source between Max and Pd is available thanks to flext

Max: http://www.cycling74.com
Pd: http://www.puredata.info
Flext: http://www.parasitaere-kapazitaeten.net/ext/flext/

=========
Licensing
=========

libtrancevibe is covered under the MIT License.

flext and np_trancevibe source code are covered under the GPL v2 License.

===========================
Trance Vibrator Information
===========================

More information about the TranceVibrator can be found at 

http://wiki.opendildonics.org/Rez_TranceVibrator

==================
Installation - Max
==================

- Put the .mxo(Mac)/.mxe(Windows) or .pd_[platform] file in a directory that Max/Pd will search for externals (Max: Options -> File Preferences -> Other Folders, Pd: Options -> Paths)
- Max: Put the .help file in the max-help directory of your Max/MSP installation

=========================
Platform Specifics Issues
=========================

-------
Windows
-------

- Windows will require the libusb-win32 filter driver to use the trancevibe external. This filter driver is available through the libtrancevibe sourceforge site.
- If the write timeout on windows is set to 0, writes will be quite flakey and will fail often. Always set it to at least 5. (The patch defaults it to 10)

-----
Linux
-----

- No Max/MSP build is available for linux as Max/MSP is not available for linux
- libusb requires either root access (i.e. running pd under sudo) or correct udev based USB permissions to run as non-root.

---
Mac
---

- Pd: libusb and libtrancevibe are required to be in /usr/local/lib since I can't get static builds working. These are included with the .dmg installer


ChangeLog
np_trancevibe ChangeLog
===========================

2008-03-12 Kyle Machulis <kyle@nonpolynomial.com>
	* Changed external name from nplabs.trancevibe to np_trancevibe due to flext naming restrictions
	* Ported to flext (http://www.parasitaere-kapazitaeten.net/ext/flext/) for cross compilation between Max and Puredata

2008-01-12 Kyle Machulis <kyle@nonpolynomial.com>
	* First release of nplabs.trancevibe Max/MSP external


max_trancevibe.c
/*
 * Implementation file for Trancevibrator Max/MSP External
 *
 * Copyright (c) 2005-2007 Kyle Machulis/Nonpolynomial Labs <kyle@nonpolynomial.com>
 *
 * More info on Nonpolynomial Labs @ http://www.nonpolynomial.com
 *
 * Sourceforge project @ http://www.sourceforge.net/projects/libtrancevibe/
 *
 * This library is covered by the LGPL, read LICENSE for details.
 */

#include "ext.h"
#include "trancevibe.h"

#define MAXSIZE 256

typedef struct _max_trancevibe
{
	struct object m_ob;
	t_atom m_args[MAXSIZE];
	trancevibe tv;
	unsigned char speed;
	unsigned int timeout;
	void *m_confirm;
} t_max_trancevibe;

void *max_trancevibe_class;

void *max_trancevibe_new(t_symbol *s, short ac, t_atom *av);

void max_trancevibe_anything(t_max_trancevibe* x, t_symbol *msg, short argc, t_atom *argv);
void max_trancevibe_speed(t_max_trancevibe* x, long t);
void max_trancevibe_assist(t_max_trancevibe *x, void *b, long m, long a, char *s);
void max_trancevibe_bang(t_max_trancevibe *x);

void main()
{
	setup((t_messlist **)&max_trancevibe_class, (method)max_trancevibe_new,0L, (short)sizeof(t_max_trancevibe), 0L, A_GIMME, 0);
	addbang((method)max_trancevibe_bang);
	addinx((method)max_trancevibe_speed,1);
	addmess((method)max_trancevibe_anything, "anything", A_GIMME, 0);
	addmess((method)max_trancevibe_assist,"assist", A_CANT, 0);
}

void *max_trancevibe_new(t_symbol *s, short ac, t_atom *av)
{
	t_max_trancevibe *x;
	x = (t_max_trancevibe *)newobject(max_trancevibe_class);
	x->speed = 0;
	x->tv = NULL;
	intin(x, 1);
#ifdef WIN32
	x->timeout = 10;
#else 
	x->timeout = 1;
#endif
	x->m_confirm = bangout(x);
	return x;
}

void max_trancevibe_anything(t_max_trancevibe* x, t_symbol *msg, short argc, t_atom *argv)
{
	if(!strcmp(msg->s_name, "open"))
	{
		int ret;
		if(argc == 1)
		{
			ret = trancevibe_open(&(x->tv), argv[0].a_w.w_long);
		}
		else
		{
			ret = trancevibe_open(&(x->tv), 0);
		}
		if(!ret)
		{
			outlet_bang(x->m_confirm);
		}
		else
		{
			post("Cannot connect to trancevibe");
		}
	}
	else if (!strcmp(msg->s_name, "count"))
	{
		post("Trancevibes Connected to System: %d", trancevibe_get_count());
	}
	else if (!strcmp(msg->s_name, "timeout"))
	{
		if(argc == 1)
		{
			x->timeout = argv[0].a_w.w_long;
		}
		else
		{
			post("Timeout requires 1 argument (timeout in milliseconds)");
		}
	}
	else if (!strcmp(msg->s_name, "close"))
	{
		if(x->tv)
		{
			trancevibe_close(x->tv);
			x->tv = NULL;
			outlet_bang(x->m_confirm);
		}
	}
	else
	{
		post("Not a valid message: %s", msg->s_name);
	}
}

void max_trancevibe_bang(t_max_trancevibe* x)
{
	if(!(x->tv))
	{
		post("Trancevibe not connected");
		return;
	}
	if(!trancevibe_set_speed(x->tv, x->speed, x->timeout)) outlet_bang(x->m_confirm);
}

void max_trancevibe_speed(t_max_trancevibe* x, long t)
{
	if(t > 255)
	{
		t = 255;
		post("Clamping trancevibe speed value to 255");
	}
	x->speed = t;
}

void max_trancevibe_assist(t_max_trancevibe *x, void *b, long m, long a, char *s)
{
	if (m==ASSIST_INLET) {
		switch (a) {
			case 0:
				sprintf(s, "Command Input");
			case 1:
				sprintf(s, "Speed Input");
		}
	}
	else {
		sprintf(s,"Bangs on successful connection");
	}
}


src/nplabs.trancevibe.c
/*
 * Implementation file for Trancevibrator Max/MSP External
 *
 * Copyright (c) 2005-2008 Kyle Machulis/Nonpolynomial Labs <kyle@nonpolynomial.com>
 *
 * More info on Nonpolynomial Labs @ http://www.nonpolynomial.com
 *
 * Sourceforge project @ http://www.sourceforge.net/projects/libtrancevibe/
 *
 * This library is covered by the MIT license, read LICENSE for details.
 */

#include "ext.h"
#include "trancevibe.h"

#define MAXSIZE 256

typedef struct _max_trancevibe
{
	struct object m_ob;
	t_atom m_args[MAXSIZE];
	trancevibe tv;
	unsigned char speed;
	unsigned int timeout;
	void *m_confirm;
} t_max_trancevibe;

void *max_trancevibe_class;

void *max_trancevibe_new(t_symbol *s, short ac, t_atom *av);

void max_trancevibe_anything(t_max_trancevibe* x, t_symbol *msg, short argc, t_atom *argv);
void max_trancevibe_speed(t_max_trancevibe* x, long t);
void max_trancevibe_assist(t_max_trancevibe *x, void *b, long m, long a, char *s);
void max_trancevibe_bang(t_max_trancevibe *x);

void main()
{
	setup((t_messlist **)&max_trancevibe_class, (method)max_trancevibe_new,0L, (short)sizeof(t_max_trancevibe), 0L, A_GIMME, 0);
	addbang((method)max_trancevibe_bang);
	addinx((method)max_trancevibe_speed,1);
	addmess((method)max_trancevibe_anything, "anything", A_GIMME, 0);
	addmess((method)max_trancevibe_assist,"assist", A_CANT, 0);
}

void *max_trancevibe_new(t_symbol *s, short ac, t_atom *av)
{
	t_max_trancevibe *x;
	x = (t_max_trancevibe *)newobject(max_trancevibe_class);
	x->speed = 0;
	x->tv = NULL;
	intin(x, 1);
#ifdef WIN32
	x->timeout = 10;
#else 
	x->timeout = 1;
#endif
	x->m_confirm = bangout(x);
	return x;
}

void max_trancevibe_anything(t_max_trancevibe* x, t_symbol *msg, short argc, t_atom *argv)
{
	if(!strcmp(msg->s_name, "open"))
	{
		int ret;
		if(argc == 1)
		{
			ret = trancevibe_open(&(x->tv), argv[0].a_w.w_long);
		}
		else
		{
			ret = trancevibe_open(&(x->tv), 0);
		}
		if(!ret)
		{
			outlet_bang(x->m_confirm);
		}
		else
		{
			post("Cannot connect to trancevibe");
		}
	}
	else if (!strcmp(msg->s_name, "count"))
	{
		post("Trancevibes Connected to System: %d", trancevibe_get_count());
	}
	else if (!strcmp(msg->s_name, "timeout"))
	{
		if(argc == 1)
		{
			x->timeout = argv[0].a_w.w_long;
		}
		else
		{
			post("Timeout requires 1 argument (timeout in milliseconds)");
		}
	}
	else if (!strcmp(msg->s_name, "close"))
	{
		if(x->tv)
		{
			if(!trancevibe_set_speed(x->tv, 0, x->timeout)) outlet_bang(x->m_confirm);
			trancevibe_close(x->tv);
			x->tv = NULL;
			outlet_bang(x->m_confirm);
		}
	}
	else
	{
		post("Not a valid message: %s", msg->s_name);
	}
}

void max_trancevibe_bang(t_max_trancevibe* x)
{
	if(!(x->tv))
	{
		post("Trancevibe not connected");
		return;
	}
	if(!trancevibe_set_speed(x->tv, x->speed, x->timeout)) outlet_bang(x->m_confirm);
}

void max_trancevibe_speed(t_max_trancevibe* x, long t)
{
	if(t > 255)
	{
		t = 255;
		post("Clamping trancevibe speed value to 255");
	}
	x->speed = t;
}

void max_trancevibe_assist(t_max_trancevibe *x, void *b, long m, long a, char *s)
{
	if (m==ASSIST_INLET) {
		switch (a) {
			case 0:
				sprintf(s, "Command Input");
			case 1:
				sprintf(s, "Speed Input");
		}
	}
	else {
		sprintf(s,"Bangs on successful connection");
	}
}


src/nplabs.trancevibe.flext.cpp
/* 
flext tutorial - simple 1 

Copyright (c) 2002,2003 Thomas Grill (xovo@gmx.net)
For information on usage and redistribution, and for a DISCLAIMER OF ALL
WARRANTIES, see the file, "license.txt," in this distribution.  

-------------------------------------------------------------------------

This is an example of a simple object doing a float inversion
*/

// include flext header
#include <flext.h>
#include <trancevibe.h>

// check for appropriate flext version
#if !defined(FLEXT_VERSION) || (FLEXT_VERSION < 400)
#error You need at least flext version 0.4.0
#endif


// define the class that stands for a pd/Max object
// Attention: the class name must be the same as the object name!! (without an eventual ~)
// Special names are possible with the usage of libraries (see the lib1 tutorial example)

class np_trancevibe:
	// inherit from basic flext class
	public flext_base
{
	// obligatory flext header (class name,base class name)
	FLEXT_HEADER(np_trancevibe,flext_base)
 
public:
	// constructor
	np_trancevibe() :
	mTranceVibe(NULL),
		mTimeout(10)
	{
		// define inlets:
		// first inlet must always be of type anything (or signal for dsp objects)
		AddInAnything("Command Input");
		
		AddInInt("Speed Input");  // add one inlet for any message

		AddOutBang("Bangs on successful connection/command");
		
		// register methods
		FLEXT_ADDMETHOD(0, trancevibe_anything);
		FLEXT_ADDMETHOD(1, trancevibe_speed);
	} 

	virtual ~np_trancevibe()
	{
		if(mTranceVibe)
		{
			close();
		}
	}
	
protected:
	trancevibe mTranceVibe;
	int mTimeout;

	void trancevibe_anything(const t_symbol *msg,int argc,t_atom *argv)
	{
		
		if(!strcmp(msg->s_name, "open"))
		{
			int ret;
			if(mTranceVibe)
			{
			}
			if(argc == 1)
			{
				post("Opening %d", GetInt(argv[0]));
				ret = trancevibe_open(&mTranceVibe, GetInt(argv[0]));
			}
			else
			{
				post("Opening default");
				ret = trancevibe_open(&mTranceVibe, 0);
			}
			if(ret >= 0)
			{
				ToOutBang(0);
			}
			else
			{
				post("Cannot connect to trancevibe");
			}
		}
		else if (!strcmp(msg->s_name, "count"))
		{
			post("Trancevibes Connected to System: %d", trancevibe_get_count());
		}
		else if (!strcmp(msg->s_name, "timeout"))
		{
			if(argc == 1)
			{
				mTimeout = GetInt(argv[0]);
			}
			else
			{
				post("Timeout requires 1 argument (timeout in milliseconds)");
			}
		}
		else if (!strcmp(msg->s_name, "close"))
		{
			close();
		}
		else
		{
			post("Not a valid message: %s", msg->s_name);
		}
	}

	void close()
	{
		if(mTranceVibe)
		{
			trancevibe_set_speed(mTranceVibe, 0, mTimeout);
			trancevibe_close(mTranceVibe);
			mTranceVibe = NULL;
		}
	}
	
	void trancevibe_speed(int input)  // method for float values
	{
		int result;
		if(!mTranceVibe)
		{
			post("Trancevibe not connected");
		}
		if(input > 255 || input < 0)
		{
			if(input > 255) input = 255;
			else if (input < 0) input = 0;
			post("Speed input must be between 0 and 255");
		}
		trancevibe_set_speed(mTranceVibe, input, mTimeout);
	}

private:
	FLEXT_CALLBACK_A(trancevibe_anything)
	FLEXT_CALLBACK_I(trancevibe_speed)
};
// instantiate the class
FLEXT_NEW("np_trancevibe", np_trancevibe)





src/np_trancevibe.flext.cpp
/* 
flext tutorial - simple 1 

Copyright (c) 2002,2003 Thomas Grill (xovo@gmx.net)
For information on usage and redistribution, and for a DISCLAIMER OF ALL
WARRANTIES, see the file, "license.txt," in this distribution.  

-------------------------------------------------------------------------

This is an example of a simple object doing a float inversion
*/

// include flext header
#include <flext.h>
#include <trancevibe.h>

// check for appropriate flext version
#if !defined(FLEXT_VERSION) || (FLEXT_VERSION < 400)
#error You need at least flext version 0.4.0
#endif


// define the class that stands for a pd/Max object
// Attention: the class name must be the same as the object name!! (without an eventual ~)
// Special names are possible with the usage of libraries (see the lib1 tutorial example)

class np_trancevibe:
	// inherit from basic flext class
	public flext_base
{
	// obligatory flext header (class name,base class name)
	FLEXT_HEADER(np_trancevibe,flext_base)
 
public:
	// constructor
	np_trancevibe() :
	mTranceVibe(NULL),
		mTimeout(10)
	{
		// define inlets:
		// first inlet must always be of type anything (or signal for dsp objects)
		AddInAnything("Command Input");
		
		AddInInt("Speed Input");  // add one inlet for any message

		AddOutBang("Bangs on successful connection/command");
		
		// register methods
		FLEXT_ADDMETHOD(0, trancevibe_anything);
		FLEXT_ADDMETHOD(1, set_speed);
	} 

	virtual ~np_trancevibe()
	{
		if(mTranceVibe)
		{
			close();
		}
	}
	
protected:
	trancevibe mTranceVibe;
	int mTimeout;
	int mSpeed;

	void trancevibe_anything(const t_symbol *msg,int argc,t_atom *argv)
	{
		
		
		if(!strcmp(msg->s_name, "open"))
		{
			int ret;
			if(mTranceVibe)
			{
				trancevibe_close(mTranceVibe);
			}
			if(argc == 1)
			{
				post("Opening %d", GetInt(argv[0]));
				ret = trancevibe_open(&mTranceVibe, GetInt(argv[0]));
			}
			else
			{
				post("Opening default");
				ret = trancevibe_open(&mTranceVibe, 0);
			}
			if(ret >= 0)
			{
				ToOutBang(0);
			}
			else
			{
				post("Cannot connect to trancevibe");
			}
		}
		else if (!strcmp(msg->s_name, "count"))
		{
			post("Trancevibes Connected to System: %d", trancevibe_get_count());
			ToOutBang(0);
		}
		else if (!strcmp(msg->s_name, "timeout"))
		{
			if(argc == 1)
			{
				mTimeout = GetInt(argv[0]);
				ToOutBang(0);
			}
			else
			{
				post("Timeout requires 1 argument (timeout in milliseconds)");
			}
		}
		else if (!strcmp(msg->s_name, "close"))
		{
			close();
			ToOutBang(0);
		}
		else if (!strcmp(msg->s_name, "bang"))
		{
			//If no argument, we consider it a bang
			if(mTranceVibe)
			{
				trancevibe_speed(mSpeed);
				ToOutBang(0);
			}
			else
			{
				post("Not connected to trancevibe");
			}
		}
		else 
		{
			post("Not a valid np_trancevibe message: %s", msg->s_name);
		}
	}

	void close()
	{
		if(mTranceVibe)
		{
			trancevibe_set_speed(mTranceVibe, 0, mTimeout);
			trancevibe_close(mTranceVibe);
			mTranceVibe = NULL;
		}
	}

	void set_speed(int speed)
	{
		mSpeed = speed;
	}
	
	void trancevibe_speed(int input)  // method for float values
	{
		int result;
		if(!mTranceVibe)
		{
			post("Trancevibe not connected");
		}
		if(input > 255 || input < 0)
		{
			if(input > 255) input = 255;
			else if (input < 0) input = 0;
			post("Speed input must be between 0 and 255");
		}
		trancevibe_set_speed(mTranceVibe, input, mTimeout);
	}

private:
	FLEXT_CALLBACK_A(trancevibe_anything)
	FLEXT_CALLBACK_I(set_speed)
};
// instantiate the class
FLEXT_NEW("np_trancevibe", np_trancevibe)



